### PR TITLE
feat: binary sort on insert

### PR DESCRIPTION
**Pull request by galvez**
@mcollina @Ethan-Arrowood @ronag :)

### CODE REVIEW

File: src/headers.js [hunk 1]
```diff
@@ -54,17 +54,24 @@ class Headers {
...
-      if (normalizedName === this[kHeaders][i]) {
-        this[kHeaders][i + 1] += `, ${normalizedValue}`
-        return
-      } else if (this[kHeaders][i] > normalizedName) {
-        index = i
-        break
+    let i = this[kHeaders].length
+    let low = -1
+    let probe
+    while (i - low > 1) {
```
Comment (Ethan-Arrowood) [inline src/headers.js:60]: can you add some comments explaining how this binary search implementation works? or if its based on some external post linking to that is sufficient

**Review by Ethan-Arrowood — CHANGES_REQUESTED** [review]
This is great; since tests pass I'm confident in binary search implementation, but I'd still like a little bit of an explanation. I've actually never used the `>>>` operator before lol

**Comment by Ethan-Arrowood** [comment]
if you could also run a write perf check against the current scan that would be appreciated.

I used this trivial script previously to evaluate map vs array and you can probably derive something from it https://gist.github.com/Ethan-Arrowood/1c79aee396eb6c2f1e78380b69ae0f0a

**Comment by galvez** [comment]
Basically, `(i + low) >>> 1` is just a way to quickly find the _middle_ and start from there.

It relies on moving the actual bits around.

You can add these links as comments:

- https://www.tbray.org/ongoing/When/200x/2003/03/22/Binary
- https://www.reddit.com/r/explainlikeimfive/comments/2yja5w/eli5_bitwise_operations/

I'll try and do the perf check some time tomorrow, gotta crash pretty soon :)

**Comment by Ethan-Arrowood** [comment]
oh perfect; that first link explains it entirely. looking forward to the benchmarks tomorrow!

**Comment by galvez** [comment]
```
write performance
Map: 0.4072289988398552ms
Array: 1.672373004257679ms
ArrayHeadersWithBinarySort: 0.21538099646568298ms

read performance
Map: 0.09028399735689163ms
Array: 0.05713299661874771ms
ArrayHeadersWithBinarySort: 0.056884996592998505ms
```

File: src/headers.js [hunk 2]
```diff
@@ -54,17 +54,24 @@ class Headers {
...
+      probe = (i + low) >>> 1
+      if (this[kHeaders][probe % 2 ? probe - 1 : probe] < normalizedName) {
```
Comment (ronag) [inline src/headers.js:62]: Not sure how smart V8 is but `& 1` might be faster than `% 2`

Comment (galvez) [reply src/headers.js:62]: @ronag weird, seems I consistently get a few ms slower for the read test using that idiom.

**Review by ronag — APPROVED** [review]

**Review by Ethan-Arrowood — APPROVED** [review]
